core/services/bulletprooftxmanager/eth_confirmer.go

```diff
@@ -140,126 +157,135 @@ func (ec *ethConfirmer) CheckForReceipts(ctx context.Context, blockNum int64) er
+	// if it were not successfully broadcast how could it possibly have a
+	// receipt?
+	//
+	// This state is reachable for example if the eth node errors so the
+	// attempt was left in_progress but the transaction was actually accepted
+	// and mined.
+	//
+	// # EthTxes update
+	// Should be self-explanatory. If we got a receipt, the eth_tx is confirmed.
+	//
+	var valueStrs []string
+	var valueArgs []interface{}
+	i := 1
+	for _, r := range receipts {
+		var receiptJSON []byte
+		receiptJSON, err = json.Marshal(r)
 		if err != nil {
-			return errors.Wrap(err, "saveReceipt failed")
-		}
-		// Conflict here shouldn't be possible because there should only ever
-		// be one receipt for an eth_tx, and if it exists then the transaction
-		// is marked confirmed which means we can never get here.
-		// However, even so, it still shouldn't be an error to re-insert a receipt we already have.
-		err = tx.Set("gorm:insert_option", "ON CONFLICT (tx_hash, block_hash) DO NOTHING").
-			Create(&models.EthReceipt{
-				Receipt:          receiptJSON,
-				TxHash:           receipt.TxHash,
-				BlockHash:        receipt.BlockHash,
-				BlockNumber:      receipt.BlockNumber.Int64(),
-				TransactionIndex: receipt.TransactionIndex,
-			}).Error
-		if err == nil || err.Error() == "sql: no rows in result set" {
-			err = multierr.Combine(
-				errors.Wrap(tx.Exec(`UPDATE eth_txes SET state = 'confirmed' WHERE id = ?`, ethTxID).Error, "saveReceipt failed to update eth_txes"),
-				// NOTE: It should always be safe to mark the attempt as
-				// broadcast here because if it were not successfully broadcast
-				// how could it possibly have a receipt?
-				//
-				// This state is reachable for example if the eth node errors
-				// so the attempt was left in_progress but the transaction was
-				// actually accepted.
-				errors.Wrap(tx.Exec(`UPDATE eth_tx_attempts SET state = 'broadcast', broadcast_before_block_num = COALESCE(eth_tx_attempts.broadcast_before_block_num, ?) WHERE hash = ?`, receipt.BlockNumber.Int64(), receipt.TxHash).Error, "saveReceipt failed to update eth_tx_attempts"),
-			)
+			return errors.Wrap(err, "saveFetchedReceipts failed to marshal JSON")
 		}
-
-		return errors.Wrap(err, "saveReceipt failed to save receipt")
-	})
+		valueStrs = append(valueStrs, fmt.Sprintf("($%v,$%v,$%v,$%v,$%v,NOW())", i, i+1, i+2, i+3, i+4))
+		valueArgs = append(valueArgs, r.TxHash, r.BlockHash, r.BlockNumber.Int64(), r.TransactionIndex, receiptJSON)
+		i += 5
+	}
+
+	/* #nosec G201 */
+	sql := `
+	WITH inserted_receipts AS (
+		INSERT INTO eth_receipts (tx_hash, block_hash, block_number, transaction_index, receipt, created_at)
+		VALUES %s
+		ON CONFLICT (tx_hash, block_hash) DO UPDATE SET
+			block_number = EXCLUDED.block_number,
+			transaction_index = EXCLUDED.transaction_index,
+			receipt = EXCLUDED.receipt
+		RETURNING eth_receipts.tx_hash, eth_receipts.block_number
+	),
+	updated_eth_tx_attempts AS (
+		UPDATE eth_tx_attempts
+		SET
+			state = 'broadcast',
```

Comment:
nit - could we use the typed strings here has args to the sprintf

Reply:
Why?

Reply:
oh didn't know these were enums in the db (thought they were strings), makes sense then